src/inc/til/hash.h

```diff
@@ -0,0 +1,149 @@
+        static constexpr size_t FNV_offset_basis = 14695981039346656037ULL;
+        static constexpr size_t FNV_prime = 1099511628211ULL;
+#else
+        static constexpr size_t FNV_offset_basis = 2166136261U;
+        static constexpr size_t FNV_prime = 16777619U;
+#endif
+
+        size_t _hash = FNV_offset_basis;
+    };
+
+    namespace details
+    {
+        template<typename T, bool enable>
+        struct conditionally_enabled_hash_trait
+        {
+            constexpr void operator()(hasher& h, const T& v) const noexcept
+            {
+                h.write(reinterpret_cast<const uint8_t*>(&v), sizeof(T));
+            }
+        };
+
+        template<typename T>
+        struct conditionally_enabled_hash_trait<T, false>
+        {
+            conditionally_enabled_hash_trait() = delete;
+            conditionally_enabled_hash_trait(const conditionally_enabled_hash_trait&) = delete;
+            conditionally_enabled_hash_trait(conditionally_enabled_hash_trait&&) = delete;
+            conditionally_enabled_hash_trait& operator=(const conditionally_enabled_hash_trait&) = delete;
+            conditionally_enabled_hash_trait& operator=(conditionally_enabled_hash_trait&&) = delete;
+        };
+    }
+
+    template<typename T>
+    struct hash_trait : details::conditionally_enabled_hash_trait<T, std::has_unique_object_representations_v<T>>
+    {
+    };
+
+    template<>
+    struct hash_trait<float>
+    {
+        constexpr void operator()(hasher& h, float v) const noexcept
+        {
+            v = v == 0.0f ? 0.0f : v; // map -0 to 0
+            h.write(reinterpret_cast<const uint8_t*>(&v), sizeof(v));
+        }
+    };
+
+    template<>
+    struct hash_trait<double>
+    {
+        constexpr void operator()(hasher& h, double v) const noexcept
+        {
+            v = v == 0.0 ? 0.0 : v; // map -0 to 0
+            h.write(reinterpret_cast<const uint8_t*>(&v), sizeof(v));
+        }
+    };
+
+    template<typename T, typename CharTraits, typename Allocator>
+    struct hash_trait<std::basic_string<T, CharTraits, Allocator>>
+    {
+        constexpr void operator()(hasher& h, const std::basic_string<T, CharTraits, Allocator>& v) const noexcept
+        {
+            h.write(reinterpret_cast<const uint8_t*>(v.data()), sizeof(T) * v.size());
+        }
+    };
+
+    template<typename T, typename CharTraits>
+    struct hash_trait<std::basic_string_view<T, CharTraits>>
+    {
+        constexpr void operator()(hasher& h, const std::basic_string_view<T, CharTraits>& v) const noexcept
+        {
+            h.write(reinterpret_cast<const uint8_t*>(v.data()), sizeof(T) * v.size());
+        }
+    };
+
+    template<typename T, typename = std::enable_if_t<std::is_integral_v<T> || std::is_enum_v<T>>>
+    constexpr size_t hash(const T v) noexcept
+    {
+        auto h = static_cast<size_t>(v);
+        if constexpr (sizeof(size_t) == 4)
+        {
+            h ^= h >> 16;
+            h *= UINT32_C(0x85ebca6b);

```

Comment:
explain source of magic constants?

Reply:
Oh, so they're just part of the public domain hash and are literally magic. Got it.